File: Difficulty: Medium/Meeting Rooms II/meeting-rooms-ii.cpp
class Solution {
  public:
    int minMeetingRooms(vector<int> &start, vector<int> &end) {
        // code here
        sort(start.begin(),start.end());
        sort(end.begin(),end.end());
        int i = 0,j = 0;
        int current_rooms = 0, max_rooms = 0;
        while(i < start.size()){
            if(start[i] < end[j]){
                current_rooms++;
                i++;
            }
            else{
                current_rooms--;
                j++;
            }
            max_rooms = max(max_rooms, current_rooms);
        }
        
        return max_rooms;
    }
};

File: Difficulty: Medium/Meeting Rooms II/meeting-rooms-ii.java
class Solution {
    public int minMeetingRooms(int[] start, int[] end) {
        // code here
        // Step 1: Dono arrays ko sort karo
        Arrays.sort(start);
        Arrays.sort(end);

        int i = 0; // Pointer for start
        int j = 0; // Pointer for end

        int currentRooms = 0;
        int maxRooms = 0;

        // Loop tab tak chalega jab tak saari meetings shuru nahi ho jaatin
        while (i < start.length) {

            // start[i] < end[j] (True) -> Naya banda aaya, Room +1
            if (start[i] < end[j]) {
                currentRooms++;
                i++;
            } 
            // start[i] < end[j] (False, jaise 8 < 7) -> Purana banda gaya, Room -1
            else {
                currentRooms--;
                j++;
            }

            // Peak rooms ka track rakho
            maxRooms = Math.max(maxRooms, currentRooms);
        }

        return maxRooms;
    }
}
